components/builder-originsrv/src/migrations/2018-10-10-198563_enum_visibility/up.sql

```diff
@@ -0,0 +1,256 @@
+    LANGUAGE sql STABLE
+    AS $$
+        SELECT COUNT(*) OVER () AS total_count, op.ident
+        FROM origin_packages op
+        WHERE op.ident LIKE (op_ident  || '%')
+        AND op.visibility = ANY(op_visibilities)
+        ORDER BY op.ident DESC
+        LIMIT op_limit
+        OFFSET op_offset;
+$$;
+
+CREATE OR REPLACE FUNCTION get_origin_packages_unique_for_origin_v5(op_origin text, op_limit bigint, op_offset bigint, op_visibilities origin_package_visibility[]) RETURNS TABLE(total_count bigint, name text)
+    LANGUAGE sql STABLE
+    AS $$
+        SELECT COUNT(*) OVER () AS total_count, op.name
+        FROM origins o INNER JOIN origin_packages op ON o.id = op.origin_id
+        WHERE o.name = op_origin
+        AND op.visibility = ANY(op_visibilities)
+        GROUP BY op.name
+        ORDER BY op.name ASC
+        LIMIT op_limit
+        OFFSET op_offset;
+$$;
+
+CREATE OR REPLACE FUNCTION insert_origin_package_v5 (
+  op_origin_id bigint,
+  op_owner_id bigint,
+  op_name text,
+  op_ident text,
+  op_checksum text,
+  op_manifest text,
+  op_config text,
+  op_target text,
+  op_deps text[],
+  op_tdeps text[],
+  op_exposes smallint[],
+  op_visibility origin_package_visibility
+) RETURNS SETOF origin_packages AS $$
+    DECLARE
+      inserted_package origin_packages;
+      channel_id bigint;
+    BEGIN
+        INSERT INTO origin_packages (origin_id, owner_id, name, ident, ident_array, checksum, manifest, config, target, deps, tdeps, exposes, visibility)
+              VALUES (op_origin_id, op_owner_id, op_name, op_ident, regexp_split_to_array(op_ident, '/'), op_checksum, op_manifest, op_config, op_target, op_deps, op_tdeps, op_exposes, op_visibility)
+              ON CONFLICT ON CONSTRAINT origin_packages_ident_key DO
+                UPDATE set checksum=op_checksum
+              RETURNING * into inserted_package;
+
+        SELECT id FROM origin_channels WHERE origin_id = op_origin_id AND name = 'unstable' INTO channel_id;
+        PERFORM promote_origin_package_v1(channel_id, inserted_package.id);
+
+        RETURN NEXT inserted_package;
+        RETURN;
+    END
+$$ LANGUAGE plpgsql VOLATILE;
+
+
+CREATE OR REPLACE FUNCTION add_audit_package_entry_v3(p_origin text, p_package text, p_channel text, p_operation origin_package_operation, p_trigger package_channel_trigger, p_requester_id bigint, p_requester_name text) RETURNS SETOF audit_package
+    LANGUAGE sql
+    AS $$
+INSERT INTO audit_package (origin_id, package_id, channel_id, operation, trigger, requester_id, requester_name)
+VALUES (
+    (SELECT id FROM origins where name = p_origin),
+    (SELECT id FROM get_origin_package_v5(p_package, '{public,private,hidden}')),
+    (SELECT id FROM get_origin_channel_v1(p_origin, p_channel)),
+    p_operation, p_trigger, p_requester_id, p_requester_name)
+RETURNING *;
+$$;
+
+CREATE OR REPLACE FUNCTION promote_origin_package_v3(in_origin text, in_ident text, to_channel text) RETURNS void
```

Comment:
Since you're creating a new promote sproc, could you modify this to be idemponent?  Currently we fail with an error 500 if we try to promote an already promoted package to a channel via the API.  Also add in a test to catch that case.